salt/pillar/__init__.py

```diff
@@ -273,83 +273,158 @@ def __init__(self, opts, grains, minion_id, saltenv, ext=None, functions=None,
         self.functions = functions
         self.pillar_override = pillar_override
         self.pillarenv = pillarenv
+        self.cache = None
+        self.ext_cache = None
 
         if saltenv is None:
             self.saltenv = 'base'
         else:
             self.saltenv = saltenv
 
         # Determine caching backend
-        self.cache = salt.utils.cache.CacheFactory.factory(
-                self.opts['pillar_cache_backend'],
-                self.opts['pillar_cache_ttl'],
-                minion_cache_path=self._minion_cache_path(minion_id))
-
-    def _minion_cache_path(self, minion_id):
+        if self.opts.get('pillar_cache', False):
+            self.cache = salt.utils.cache.CacheFactory.factory(
+                    self.opts['pillar_cache_backend'],
+                    self.opts['pillar_cache_ttl'],
+                    minion_cache_path=self._minion_cache_path('pillar_cache', minion_id))
+
+        if self.opts.get('ext_pillar_cache', False):
+            self.ext_cache = salt.utils.cache.CacheFactory.factory(
+                    self.opts['ext_pillar_cache_backend'],
+                    self.opts['ext_pillar_cache_ttl'],
+                    minion_cache_path=self._minion_cache_path('ext_pillar_cache', minion_id))
+
+    def _minion_cache_path(self, cache_type, minion_id):
         '''
         Return the path to the cache file for the minion.
 
         Used only for disk-based backends
         '''
-        return os.path.join(self.opts['cachedir'], 'pillar_cache', minion_id)
+        return os.path.join(self.opts['cachedir'], cache_type, minion_id)
 
-    def fetch_pillar(self):
+    def fetch_pillar(self, fetch_pillar=True, fetch_ext_pillar=True):
```

Comment:
It's not clear what these do and should probably be documented in the doc string below.